XMonad/Util/Parser.hs

```diff
@@ -0,0 +1,215 @@
+
+>>> runParser pCombination' "<f>"
+Just "f"
+
+One might also define @pLangle'@ as @string "<" <* eof@, which would
+enable a definition of @pCombination' = pLangle' <> pLongerSequence'@.
+
+For example uses, see "XMonad.Util.EZConfig" or "XMonad.Prompt.OrgMode".
+-}
+
+-- Parser :: Type -> Type
+newtype Parser a = Parser (ReadP a)
+  deriving newtype (Functor, Applicative, Monad)
+
+instance Semigroup (Parser a) where
+  -- | Local, exclusive, left-biased choice: If left parser locally
+  -- produces any result at all, then right parser is not used.
+  (<>) :: Parser a -> Parser a -> Parser a
+  (<>) = coerce ((<++) @a)
+
+instance Monoid (Parser a) where
+  -- | A parser that always fails.
+  mempty :: Parser a
+  mempty = Parser empty
+
+instance Alternative Parser where
+  empty :: Parser a
+  empty = mempty
+
+  (<|>) :: Parser a -> Parser a -> Parser a
+  (<|>) = (<>)
+
+-- | Run a parser on a given string.
+runParser :: Parser a -> String -> Maybe a
+runParser (Parser p) = fmap fst . listToMaybe . ReadP.readP_to_S p
+
+-- | Consume and return the next character.  Fails if there is no input
+-- left.
+get :: Parser Char
+get = coerce ReadP.get
+
+-- | Look-ahead: return the part of the input that is left, without
+-- consuming it.
+look :: Parser String
+look = coerce ReadP.look
+
+-- | Succeeds if and only if we are at the end of input.
+eof :: Parser ()
+eof = coerce ReadP.eof
+
+-- | Parse an integral number number.
+num :: (Read a, Integral a) => Parser a
+num = read <$> munch1 isDigit
+{-# SPECIALISE num :: Parser Word    #-}
+{-# SPECIALISE num :: Parser Int     #-}
+{-# SPECIALISE num :: Parser Integer #-}
+
+-- | Parse and return the specified character.
+char :: Char -> Parser Char
+char = coerce ReadP.char
+
+-- | Parse and return the specified string.
+string :: String -> Parser String
+string = coerce ReadP.string
+
+-- | Skip all whitespace.
+skipSpaces :: Parser ()
+skipSpaces = coerce ReadP.skipSpaces
+
+-- | Consume and return the next character if it satisfies the specified
+-- predicate.
+satisfy :: (Char -> Bool) -> Parser Char
+satisfy = coerce ReadP.satisfy
+
+-- | Combine all parsers in the given list in a left-biased way.
+choice :: [Parser a] -> Parser a
+choice = foldl' (<>) mempty
+
+-- | Parse the first zero or more characters satisfying the predicate.
+-- Always succeeds, exactly once having consumed all the characters.
+-- Hence, this is __not__ the same as @many (satisfy p)@.
+munch :: (Char -> Bool) -> Parser String
+munch = coerce ReadP.munch
+
+-- | Parse the first one or more characters satisfying the predicate.
+-- Fails if none, else succeeds exactly once having consumed all the
+-- characters.  Hence, this is __not__ the same as @many1 (satisfy p)@.
+munch1 :: (Char -> Bool) -> Parser String
+munch1 = coerce ReadP.munch1
+
+-- | @endBy p sep@ parses zero or more occurrences of @p@, separated and
+-- ended by @sep@.
+endBy :: ReadP a -> ReadP sep -> ReadP [a]
```

Comment:
Any reason this isn't `Parser a -> Parser sep -> Parser [a]`?

Reply:
There is a perfectly valid reason—I'm a dunce :)